presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateMostFrequentHistogram.java

```diff
@@ -0,0 +1,139 @@
+    private int capacity;
+    private ApproximateMostFrequentBucketSerializer<K> serializer;
+    private ApproximateMostFrequentBucketDeserializer<K> deserializer;
+
+    /**
+     * @param maxBuckets The maximum number of elements stored in the bucket.
+     * @param capacity The maximum capacity of the stream summary data structure.
+     * @param serializer It serializes a bucket into varbinary slice.
+     * @param deserializer It appends a bucket into the histogram.
+     */
+    public ApproximateMostFrequentHistogram(int maxBuckets,
+                                            int capacity,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        requireNonNull(serializer, "serializer is null");
+        requireNonNull(deserializer, "deserializer is null");
+        streamSummary = new StreamSummary<>(capacity);
+        this.maxBuckets = maxBuckets;
+        this.capacity = capacity;
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+    }
+
+    public ApproximateMostFrequentHistogram(Slice serialized,
+                                            ApproximateMostFrequentBucketSerializer<K> serializer,
+                                            ApproximateMostFrequentBucketDeserializer<K> deserializer)
+    {
+        SliceInput input = serialized.getInput();
+
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+
+        this.maxBuckets = input.readInt();
+        this.capacity = input.readInt();
+        this.streamSummary = new StreamSummary<>(capacity);
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+
+        for (int i = 0; i < maxBuckets; i++) {
+            this.deserializer.deserialize(input, this);
+        }
+    }
+
+    public void add(K value)
+    {
+        streamSummary.offer(value);
+    }
+
+    public void add(K value, long incrementCount)
+    {
+        streamSummary.offer(value, toIntExact(incrementCount));
+    }
+
+    public Slice serialize()
+    {
+        DynamicSliceOutput output = new DynamicSliceOutput(1000);
+        List<Counter<K>> counters = streamSummary.topK(maxBuckets);
+        output.appendByte(FORMAT_TAG);
+        output.appendInt(maxBuckets);
+        output.appendInt(capacity);
+        // Serialize key and counts.
+        for (Counter<K> counter : counters) {
+            serializer.serialize(counter.getItem(), counter.getCount(), output);
+        }
+
+        return output.getUnderlyingSlice();
+    }
+
+    public void merge(ApproximateMostFrequentHistogram<K> other)
+    {
+        List<Counter<K>> counters = other.streamSummary.topK(maxBuckets);
+        for (Counter<K> counter : counters) {
+            add(counter.getItem(), counter.getCount());
+        }
+    }
+
+    public Map<K, Long> getBuckets()
```

Comment:
This could return an object with two parallel arrays or lists, one with the keys and one for the values (e.g., `long[]`) to avoid having to create a map every time. None of the callers really needs the map, since they iterate over the entries and write them out to the output.

Alternatively, it could be replaced with a method that walks over the entires and calls back with each entry. E.g.:

```java
forEachBucket((key, value) -> {
    VarcharType.VARCHAR.writeSlice(entryBuilder, key));
    BigintType.BIGINT.writeLong(entryBuilder, value);
});
```